Spigot-API-Patches/0058-Profile-Lookup-Events.patch

```diff
@@ -0,0 +1,221 @@
++            <scope>compile</scope> <!-- expose to Plugins -->
++        </dependency>
+         <dependency>
+             <groupId>co.aikar</groupId>
+             <artifactId>fastutil-lite</artifactId>
+diff --git a/src/main/java/com/destroystokyo/paper/event/profile/LookupProfileEvent.java b/src/main/java/com/destroystokyo/paper/event/profile/LookupProfileEvent.java
+new file mode 100644
+index 00000000..fddc2015
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/profile/LookupProfileEvent.java
+@@ -0,0 +1,32 @@
++package com.destroystokyo.paper.event.profile;
++
++import com.mojang.authlib.GameProfile;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++
++/**
++ * Allows a plugin to be notified anytime AFTER a Profile has been looked up from the Mojang API
++ *
++ * This is an opportunity to view the response and potentially cache things.
++ */
++public class LookupProfileEvent extends Event {
++    private final GameProfile profile;
++
++    public LookupProfileEvent(GameProfile profile) {
++        this.profile = profile;
++    }
++
++    public GameProfile getProfile() {
++        return profile;
++    }
++
++    private static final HandlerList handlers = new HandlerList();
++
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java b/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
+new file mode 100644
+index 00000000..94928526
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/profile/PreLookupProfileEvent.java
+@@ -0,0 +1,148 @@
++package com.destroystokyo.paper.event.profile;
++
++import com.google.common.collect.Multimap;
++import com.google.common.collect.Sets;
++import com.mojang.authlib.GameProfile;
++import com.mojang.authlib.GameProfileRepository;
++import com.mojang.authlib.ProfileLookupCallback;
++import com.mojang.authlib.properties.Property;
++import com.mojang.authlib.properties.PropertyMap;
++import org.bukkit.Bukkit;
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++
++import javax.annotation.Nonnull;
++import javax.annotation.Nullable;
++import java.util.Set;
++import java.util.UUID;
++
++/**
++ * Allows a plugin to intercept a Profile Lookup for a Profile by name
++ *
++ * At the point of event fire, the UUID and properties are unset.
++ *
++ * If a plugin sets the UUID, and optionally the properties, the API call to look up the profile may be skipped.
++ *
++ */
++public class PreLookupProfileEvent extends Event {
++    private final String name;
++
++    private UUID uuid;
++    private PropertyMap properties = new PropertyMap();
++
++    public PreLookupProfileEvent(@Nonnull String name) {
++        super(!Bukkit.isPrimaryThread());
```

Comment:
This is standard Bukkit stuff. This event has no guarantee of its async status, so plugins should check event.isAsync()